spec/memoize_query_spec.rb

```diff
@@ -612,4 +537,364 @@ def expect_to_eq_with_or_without_redis
+    RSpec.shared_examples 'memoizes queries with both comparator and other bound conditions' do
+      it 'memoizes queries with both comparator and other bound conditions' do
+        if defined?(relation1_with_comparator_and_other)
+          expect_to_use_redis do
+            expect(relation1_with_comparator_and_other.to_a).to eq(result)
+          end
+        end
+  
+        if defined?(relation2_with_comparator_and_other)
+          expect_to_use_redis do
+            expect(relation2_with_comparator_and_other.to_a).to eq(result)
+          end
+        end
+      end
+    end
+
+    RSpec.shared_examples 'when update happens to the records' do
+      it 'it updates the affected query result' do
+        update
+        if defined?(relation1_with_comparator_and_other)
+          expect_to_use_redis do
+            expect(relation1_with_comparator_and_other.reload.to_a).to eq(result)
+          end
+        end
+  
+        if defined?(relation2_with_comparator_and_other)
+          expect_to_use_redis do
+            expect(relation2_with_comparator_and_other.reload.to_a).to eq(result)
+          end
+        end
+      end
+    end
+
+    RSpec.shared_examples 'with local cache' do
+      it 'only invalidates the affected query result sets' do
+        RedisMemo::Cache.with_local_cache do 
+          if defined?(relation1_with_comparator_and_other)
+            expect_to_use_redis do
+              expect(relation1_with_comparator_and_other.to_a).to eq(unaffected_result)
+            end
+          end
+  
+          if defined?(relation2_with_comparator_and_other)
+            expect_to_use_redis do
+              expect(relation2_with_comparator_and_other.to_a).to eq(unaffected_result)
+            end
+          end
+  
+          unaffected_creation
+  
+          # Does not affect with the creation because it is not dependent
+          if defined?(relation1_with_comparator_and_other)
+            expect_not_to_use_redis do
+              expect(relation1_with_comparator_and_other.to_a).to eq(unaffected_result)
+            end
+          end
+  
+          if defined?(relation2_with_comparator_and_other)
+            expect_not_to_use_redis do
+              expect(relation2_with_comparator_and_other.to_a).to eq(unaffected_result)
+            end
+          end
+  
+          update
+  
+          # Will affect with the update
+          if defined?(relation1_with_comparator_and_other)
+            expect_to_use_redis do
+              expect(relation1_with_comparator_and_other.reload.to_a).to eq(affected_result)
+            end
+          end
+  
+          if defined?(relation2_with_comparator_and_other)
+            expect_to_use_redis do
+              expect(relation2_with_comparator_and_other.reload.to_a).to eq(affected_result)
+            end
+          end
+        end
+      end
+    end
+  end
+
+  context 'when the queries have greater than operators (unbound)' do
```

Comment:
Also, these tests mostly look the same, the only difference being the operator `<, <=, >, >=`, so it looks like there's a lot of code duplication. I feel like it'd be sufficient to just have a list of 1. relations with only comparators, and  2. relations with comparator and bound conditions, and combine most of these test cases together.

Reply:
> This test doesn't contain only unbound conditions right?

and that actually means "contain unbound", I updated the title though

> 1. relations with only comparators, and 2. relations with comparator and bound conditions

updated in each context

>  combine most of these test cases together

I combined the cases for > with >= and < with <=, didn't combine all of them into one since there is nuance in the update and create action (used to test the results are valid after update) for different operators. I was able to resolve the code dup. For whether we should test all cases, i still think that we should test each test case as possible, to everyone sure it works correctly. I think it does not matter too much to just test one more relation.